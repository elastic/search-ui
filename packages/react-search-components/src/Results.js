import PropTypes from "prop-types";
import React from "react";

function Results({ children }) {
  return <ul className="sui-results">{children}</ul>;
}

Results.propTypes = {
  children: PropTypes.node.isRequired
};

export default Results;
