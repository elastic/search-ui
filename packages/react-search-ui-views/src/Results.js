import PropTypes from "prop-types";
import React from "react";

import { appendClassName } from "./view-helpers";

function Results({ children, className, ...rest }) {
  return (
    <ul
      className={appendClassName("sui-results-container", className)}
      {...rest}
    >
      {children}
    </ul>
  );
}

Results.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

export default Results;
