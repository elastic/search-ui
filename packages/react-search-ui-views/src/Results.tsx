import type { SearchContextState, SearchResult } from "@elastic/search-ui";
import React from "react";
import { BaseContainerProps } from "./types";
import { ResultViewProps } from "./Result";

import { appendClassName } from "./view-helpers";

export type ResultsContainerContext = Pick<SearchContextState, "results">;

export type ResultsContainerProps = BaseContainerProps &
  ResultsContainerContext & {
    view?: React.ComponentType<ResultsViewProps>;
    resultView?: React.ComponentType<ResultViewProps>;
    clickThroughTags?: string[];
    titleField?: string;
    urlField?: string;
    thumbnailField?: string;
    results: SearchResult[];
    shouldTrackClickThrough?: boolean;
  };

export type ResultsViewProps = BaseContainerProps;

function Results({
  children,
  className,
  ...rest
}: ResultsViewProps & React.HTMLAttributes<HTMLUListElement>) {
  return (
    <ul
      className={appendClassName("sui-results-container", className)}
      {...rest}
    >
      {children}
    </ul>
  );
}

export default Results;
