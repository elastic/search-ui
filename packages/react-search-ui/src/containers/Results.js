import PropTypes from "prop-types";
import React, { Component } from "react";
import { Result, Results } from "@elastic/react-search-ui-views";

import { withSearch } from "..";
import { Result as ResultContainer } from ".";
import { Result as ResultType } from "../types";

function getRaw(result, value) {
  if (!result[value] || !result[value].raw) return;
  return result[value].raw;
}

export class ResultsContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    clickThroughTags: PropTypes.arrayOf(PropTypes.string),
    resultView: PropTypes.func,
    titleField: PropTypes.string,
    urlField: PropTypes.string,
    thumbnailField: PropTypes.string,
    view: PropTypes.func,
    shouldTrackClickThrough: PropTypes.bool,
    // State
    results: PropTypes.arrayOf(ResultType).isRequired
  };

  static defaultProps = {
    clickThroughTags: [],
    shouldTrackClickThrough: true
  };

  render() {
    const {
      className,
      clickThroughTags,
      resultView,
      results,
      shouldTrackClickThrough,
      titleField,
      urlField,
      thumbnailField,
      view,
      ...rest
    } = this.props;

    const View = view || Results;
    const ResultView = resultView || Result;

    return View({
      className: className,
      children: results.map(result => (
        <ResultContainer
          key={`result-${getRaw(result, "id")}`}
          titleField={titleField}
          urlField={urlField}
          thumbnailField={thumbnailField}
          view={ResultView}
          result={result}
          shouldTrackClickThrough={shouldTrackClickThrough}
          clickThroughTags={clickThroughTags}
        />
      )),
      ...rest
    });
  }
}

export default withSearch(({ results }) => ({ results }))(ResultsContainer);
