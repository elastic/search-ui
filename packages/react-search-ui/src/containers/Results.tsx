import React, { Component } from "react";
import {
  Result,
  Results,
  ResultsContainerProps,
  ResultsContainerContext
} from "@elastic/react-search-ui-views";

import { withSearch } from "..";
import { Result as ResultContainer } from ".";

function getRaw(result, value) {
  if (!result[value] || !result[value].raw) return;
  return result[value].raw;
}

export class ResultsContainer extends Component<ResultsContainerProps> {
  static defaultProps = {
    clickThroughTags: [],
    shouldTrackClickThrough: true
  };

  render() {
    const {
      className,
      clickThroughTags,
      resultView,
      results,
      shouldTrackClickThrough,
      titleField,
      urlField,
      thumbnailField,
      view,
      ...rest
    } = this.props;

    const View = view || Results;
    const ResultView = resultView || Result;

    const children = results.map((result) => (
      <ResultContainer
        key={`result-${getRaw(result, "id")}`}
        titleField={titleField}
        urlField={urlField}
        thumbnailField={thumbnailField}
        view={ResultView}
        shouldTrackClickThrough={shouldTrackClickThrough}
        clickThroughTags={clickThroughTags}
        result={result}
      />
    ));

    const viewProps = {
      className,
      children,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<ResultsContainerProps, ResultsContainerContext>(
  ({ results }) => ({ results })
)(ResultsContainer);
