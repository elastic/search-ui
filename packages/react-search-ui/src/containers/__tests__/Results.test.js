import React from "react";
import { shallow } from "enzyme";
import { ResultsContainer } from "../Results";

const params = {
  results: [
    {
      id: {
        raw: "id",
        snippet: "<em>id</em>"
      },
      title: {
        raw: "title",
        snippet: "<em>title</em>"
      },
      url: {
        raw: "url",
        snippet: "<em>url</em>"
      }
    },
    {
      id: {
        raw: "id",
        snippet: "<em>id</em>"
      },
      title: {
        raw: "title",
        snippet: "<em>title</em>"
      },
      url: {
        raw: "url",
        snippet: "<em>url</em>"
      }
    }
  ],
  titleField: "title",
  urlField: "url"
};

it("supports a render prop", () => {
  // eslint-disable-next-line react/prop-types
  const render = ({ children }) => {
    return <div>{children}</div>;
  };
  const wrapper = shallow(<ResultsContainer {...params} view={render} />);
  expect(wrapper).toMatchSnapshot();
});

it("passes through props to individual Result items", () => {
  // eslint-disable-next-line react/prop-types
  const renderResult = ({ title }) => {
    return <li>{title}</li>;
  };
  const wrapper = shallow(
    <ResultsContainer
      {...params}
      renderResult={renderResult}
      shouldTrackClickThrough={true}
      clickThroughTags={["whatever"]}
    />
  );
  expect(wrapper.find("WithSearch").map(n => n.props())).toEqual([
    {
      result: {
        id: {
          raw: "id",
          snippet: "<em>id</em>"
        },
        title: {
          raw: "title",
          snippet: "<em>title</em>"
        },
        url: {
          raw: "url",
          snippet: "<em>url</em>"
        }
      },
      titleField: "title",
      urlField: "url",
      shouldTrackClickThrough: true,
      clickThroughTags: ["whatever"],
      view: renderResult
    },
    {
      result: {
        id: {
          raw: "id",
          snippet: "<em>id</em>"
        },
        title: {
          raw: "title",
          snippet: "<em>title</em>"
        },
        url: {
          raw: "url",
          snippet: "<em>url</em>"
        }
      },
      titleField: "title",
      urlField: "url",
      shouldTrackClickThrough: true,
      clickThroughTags: ["whatever"],
      view: renderResult
    }
  ]);
});
