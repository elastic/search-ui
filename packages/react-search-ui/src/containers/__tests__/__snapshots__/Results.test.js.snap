// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports a render prop 1`] = `
<div>
  <WithSearch
    clickThroughTags={Array []}
    key="result-id"
    result={
      Object {
        "id": Object {
          "raw": "id",
          "snippet": "<em>id</em>",
        },
        "title": Object {
          "raw": "title",
          "snippet": "<em>title</em>",
        },
        "url": Object {
          "raw": "url",
          "snippet": "<em>url</em>",
        },
      }
    }
    shouldTrackClickThrough={true}
    titleField="title"
    urlField="url"
    view={[Function]}
  />
  <WithSearch
    clickThroughTags={Array []}
    key="result-id"
    result={
      Object {
        "id": Object {
          "raw": "id",
          "snippet": "<em>id</em>",
        },
        "title": Object {
          "raw": "title",
          "snippet": "<em>title</em>",
        },
        "url": Object {
          "raw": "url",
          "snippet": "<em>url</em>",
        },
      }
    }
    shouldTrackClickThrough={true}
    titleField="title"
    urlField="url"
    view={[Function]}
  />
</div>
`;
